Limit test utils to a single environment (browser / ssr).

MMMalik opened the pull request:
This PR achieves the following:

1. Refactors various test utils so that tests are run for a requested browser (or ssr) rather than for the whole permuatation of browsers and React versions. Most notably `react-tester` script is parametrized and accepts a few arguments (React version, browser, flag for dry run). Config for Karma is changed to use few BrowserStack-specific adjustments.

2. Travis CI is replaced with GitHub workflows. There are 2 files for GH actions: `pr-check` is supposed to run on PRs to master only and is supposed to be a quick feedback (runs tests for a single React version and on a single browser). `master-check` is supposed to run on commits to `master` and it tests whole permutation of React versions and browsers. **As an exception and for the convenience of reviewer, `master-check` will run for this PR. A small adjustment must be committed before merging!**

3. Jobs for various browsers are run in parallel rather than sequentially. This reduces the amount of time spent on testing and allows for a better overview of running / failing jobs.
 
4. BrowserStack connectivity issues are mitigated with BS-specific settings for Karma. `karma-browserstack-launcher` is set to v1.4 which seems to be much more stable than its newer versions as per [this issue](https://github.com/karma-runner/karma-browserstack-launcher/issues/152).

After these changes there are following scripts in `package.json` that are related to tests:

- `test:ssr` - Tests in SSR mode.
- `test:dry` - A new script. It runs `react-tester` in "dry" mode. `react-tests` temp folder is NOT created and currently intalled version of React is used. It runs on Chrome by default.
- `test:browser` - Convenience wrapper for `karma start`.
- `test:all` - Runs `react-tester` for all versions of React and on a single browser (Chrome by default).
- `test` - Convenience script for running `test:dry` and linter.

On a sidenote, I tried to get rid of BS altogether, but I was stuck at getting `macos` / `windows` images to work with our setup. Installed binaries of browsers (Safari and Edge respectively) were not recognized by Karma. I didn't want to spend more time investigating it.

Closes #191.
Closes #192.

f1ames commented on .travis.yml:
```diff
@@ -1,40 +0,0 @@
-os: linux
```
:wave:

f1ames commented on karma.conf.js:
```diff
@@ -109,15 +112,24 @@ module.exports = function( config ) {
+			build: process.env.BUILD_SLUG,
+			project: 'ckeditor4',
+			video: false,
+			// (#191)
+			// This is an undocumented option of karma-browserstack-launcher. It helps to mitigate rate limiting on BrowserStack end.
+			// https://github.com/mui-org/material-ui/pull/25049
+			pollingTimeout: 10000
```
Interesting find :+1:

f1ames commented on package.json:
```diff
@@ -15,11 +15,12 @@
-		"test": "npm run test:browser && npm run test:ssr",
+		"test": "npm run test:dry",
+		"test:all": "node ./scripts/react-tester.js --react all",
 		"test:browser": "karma start",
```
I wonder if `test:browser` still makes sense, since from this PR we use react-tester as an entry point for all test runs :thinking:

MMMalik replied:
It's used from within `react-tester.js` as `npm` command.

f1ames replied:
Ahhh, somehow missed it :+1:

f1ames commented on .github/workflows/master-check.yml:
```diff
@@ -0,0 +1,56 @@
+name: Check commits to master
+
+on:
+    pull_request:
+        branches:
+            - master
```
This eventually will be triggered on commits to `master` branch, two things I guess we could improve:

*   Would it be possible to detect `[skip ci]` in the commit message which worked with Travis and we are used to using it (e.g. when updating changelog). So we can skip test runs for some commits.
*   Apart from master we could add `stable` branch and `vA.B.C` tags too.

WDYT?

MMMalik replied:
Yeah, that makes sense. It's already supported by GH actions -> see [here](https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/). I was thinking about adding `paths` filter so only changes to certain directories would trigger pipline but that requires additional maintenance of a whitelist.

So my suggestion is to try to leverage [this feature](https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/) and see if that's enough.

In addition, I will add triggers for `stable` branch and version tags.

MMMalik replied:
BTW, `paths` are described [here](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths).

f1ames replied:
We can go with `[skip ci]` for now, I see it works out-of-the-box :+1:

The `paths` thing is an interesting one. I think we could benefit from using it too (not having to remember about `skip ci` is one think less to remember about). @MMMalik could you extract an issue for that and maybe we can have it in v2? It will be easier to use if we will be having it in mind from the very beginning when working on test infrastructure. WDYT?

MMMalik replied:
Added task #197

f1ames commented on scripts/react-tester.js:
```diff
@@ -4,30 +4,46 @@ const { execSync } = require( 'child_process' );
+ * --browser <name>   Specifies environment to test.
+ *                    Possible values: 'Chrome', 'Firefox', 'BrowserStack_Safari', 'BrowserStack_Edge', 'SSR'. Defaults to: 'Chrome'.
+ * --react <version>  Specifies react version to test. Possible values: 'all', 'current' or specific version. Defaults to: 'current'.
+ * --dry              Runs tests for currently installed version of React without creating temp dir. Use for quick feedback.
```
I wonder if `--dry` is not a little misleading name here :thinking: It is usually used for a test run (haha, coincidence? :smile:). But on a serious note, maybe something like `--single` or `--single-run` or sth similar?

And now I wonder, isn't `--dry` kind of duplicating `node ./scripts/react-tester.js` run? Since by default Chrome and current React version is used. So `--dry` only prevents the script from creating tmp dir? How much time we can save thanks to this? :thinking:

f1ames replied:
And it also means that `--dry` ignores `--react` arg which could be also mentioned in the docs above (until we decide to change the way it works)?

MMMalik replied:
The idea here is to be able to run tests through `react-tester` but with locally installed `react`. It's much faster and perfect for local development.

Indeed, `dry` is not a perfect choice here although it was supposed to imply that there are no side-effects (no new directory is created, no files are removed, copied, etc.).

Now I'm thinking that this whole `dry` run can be removed in favor of simply using `test:browser` script for local development.

f1ames commented on scripts/react-tester.js:
```diff
@@ -37,52 +53,66 @@ try {
 
+/**
+ * Gets list of available React versions from npm.
+ * @returns {undefined}
```
```suggestion
 * @returns {string[]}
```

?

f1ames replied:
We usually keep one empty line between method description and the rest, like:

```javascript
/**
 * Gets list of available React versions from npm.
 *
 * @returns {undefined}
 */
```

Could you adjust the docs in this file to follow this convention?

MMMalik replied:
Okay!

f1ames commented on scripts/react-tester.js:
```diff
@@ -37,52 +53,66 @@ try {
 
+/**
+ * Gets peered version range from `package.json`.
+ * @param {*} packageInfo contents of `package.json`
```
Object or something? :thinking:

MMMalik replied:
Agreed

f1ames commented on scripts/react-tester.js:
```diff
@@ -110,41 +146,69 @@ function isLatestPatch( index, array ) {
+ * Executes npm command.
+ * @param {string} command command to execute
+ * @param {string} cwd dir where to execute command
+ * @returns {undefined}
```
Returns string or Buffer, see https://nodejs.org/api/child_process.html#child_process_child_process_execsync_command_options.

MMMalik replied:
Okay!

f1ames reviewed (changes requested):
Looking good :+1: 

I was wondering if two separate workflow files makes sense, but I see it keeps them much simpler than putting everything into a single file :+1: 

Also good job on adding docs to React tester too :+1: 

And yes, we were considering migrating from Travis to GH Actions for some time so another :+1: 

---

Could we also cover IE11 here? We state in the README it's supported (and it is AFAIR) but somehow we have missed adding it to the CI :see\_no\_evil: :sweat\_smile: 

---

Do you think it would make any sense to run additional browsers as a PR check? Or we should trust 3rd-parties (so React and CKEditor 4) simply and assume if it works on Chrome there is 99,9% chance it will work in different browsers too?

Especially, I'm thinking about IE11 which kind of lags behind modern browsers. WDYT?

---

I know it's a detail but maybe we could simplify job names, now we have:

![](https://user-images.githubusercontent.com/1061942/114700044-51fdc800-9d21-11eb-8416-7454666c6258.png)

Maybe we could get rid of the first part ("Check...") and then run time will be also visible? Like:

![](https://user-images.githubusercontent.com/1061942/114700251-938e7300-9d21-11eb-82ac-ec53aac84350.png)

Alternatively - `Run full test suite (Chrome, master)`, `Run tests (Chrome)`.

I know it's a combination of workflow and job name so maybe be hard to come up with something reasonable :thinking: WDYT?

MMMalik commented:
Thanks @f1ames for the review. 

I will look into possibility of testing IE11 through BrowserStack.

Running a full suite of tests for each PR is IMO an overkill and I would test all browsers only for pushes to `master`, `stable`, and version tags. Let's see how that goes. If we notice that browser-specific bugs are leaking into those branches too often then we can use a stricter approach.

MMMalik commented:
Okay, so I've added some changes:

- Fixed `jsdoc` formatting
- Removed `dry` completely. For local development use `npm run test` or `npm run test:browser` (it will bypass `react-tester`)
- Adjusted workflow names. Now they are more compact
- Support for IE11. This was not that trivial: changes to webpack config in `karma.config.js` were required (add polyfill, transpile all dependencies). Also, [one test](https://github.com/ckeditor/ckeditor4-react/pull/195/files#diff-adab44d193b2306bbb7ee8100ea2d4cac7fee6bebf1d6a6df9a0b3fd20cf00cdR269) was adjusted because it was failing on IE11 (`afterEach` hook was failing due to some sort of race condition)

@f1ames Please check!

f1ames commented on scripts/react-tester.js:
```diff
@@ -4,30 +4,40 @@ const { execSync } = require( 'child_process' );
+ * Commands:
+ *
+ * --browser <name>   Specifies environment to test.
+ *                    Possible values: 'Chrome', 'Firefox', 'BrowserStack_Safari', 'BrowserStack_Edge', 'SSR'. Defaults to: 'Chrome'.
```
```suggestion
 *                    Possible values: 'Chrome', 'Firefox', 'BrowserStack_Safari', 'BrowserStack_Edge', 'BrowserStack_IE11', 'SSR'. Defaults to: 'Chrome'.
```

f1ames commented on .github/workflows/test-all.yml:
```diff
@@ -0,0 +1,60 @@
+name: Test browsers
+
+on:
+    pull_request:
```
So this needs to be replaced with `push` before merge AFAIU? I think we can already change it since we are almost done :+1:

MMMalik replied:
Adjusted trigger conditions

f1ames reviewed (changes requested):
Looking good, only some minor polishing here and there and we are good to go :+1: 

> Running a full suite of tests for each PR is IMO an overkill and I would test all browsers only for pushes to `master`, `stable`, and version tags. Let's see how that goes. If we notice that browser-specific bugs are leaking into those branches too often then we can use a stricter approach.

Agree :+1: 

> * Support for IE11. This was not that trivial: changes to webpack config in `karma.config.js` were required (add polyfill, transpile all dependencies). Also, [one test](https://github.com/ckeditor/ckeditor4-react/pull/195/files#diff-adab44d193b2306bbb7ee8100ea2d4cac7fee6bebf1d6a6df9a0b3fd20cf00cdR269) was adjusted because it was failing on IE11 (`afterEach` hook was failing due to some sort of race condition)

Good job :+1: 

> * Removed `dry` completely. For local development use `npm run test` or `npm run test:browser` (it will bypass `react-tester`)

Could you adjust README slightly to mention this (and maybe very short info what other testing commands does)?

MMMalik commented:
To summarize latest changes:

- Adjusted trigger conditions as discussed: trigger full test suite on changes to `master`, `major`, `stable` and new `v*` tags.
- Adjusted README
- Added task #197

f1ames reviewed (approved):
Nothing to add here, look good :+1: Good job :+1: :tada:

f1ames commented:
Rebased onto latest `master`.
